components/skeleton/nz-skeleton.component.ts

```diff
@@ -0,0 +1,125 @@
+import { Component, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
+import { NzSkeletonAvatar, NzSkeletonParagraph, NzSkeletonTitle } from './nz-skeleton.type';
+
+@Component({
+  selector: 'nz-skeleton',
+  templateUrl: './nz-skeleton.component.html',
+  host: {
+    '[class.ant-skeleton]': 'true',
+    '[class.ant-skeleton-with-avatar]': '!!nzAvatar',
+    '[class.ant-skeleton-active]': 'nzActive'
+  }
+})
+export class NzSkeletonComponent implements OnInit, OnChanges {
+  title: NzSkeletonTitle;
+  avatar: NzSkeletonAvatar;
+  paragraph: NzSkeletonParagraph;
+  avatarClassMap;
+  rowsList: number[] = [];
+  widthList: Array<number | string> = [];
+
+  @Input() nzActive = false;
+  @Input() nzLoading = true;
+  @Input() nzTitle: NzSkeletonTitle | boolean = true;
+  @Input() nzAvatar: NzSkeletonAvatar | boolean = false;
+  @Input() nzParagraph: NzSkeletonParagraph | boolean = true;
+
+  getTitleBasicProps(hasAvatar: boolean, hasParagraph: boolean): NzSkeletonTitle {
+    if (!hasAvatar && hasParagraph) {
+      return { width: '38%' };
+    }
+    if (hasAvatar && hasParagraph) {
+      return { width: '50%' };
+    }
+    return { width: '100%' };
+  }
+
+  getAvatarBasicProps(hasTitle: boolean, hasParagraph: boolean): NzSkeletonAvatar {
+    if (hasTitle && !hasParagraph) {
+      return { shape: 'square' };
+    }
+    return { shape: 'circle' };
+  }
+
+  getParagraphBasicProps(hasAvatar: boolean, hasTitle: boolean): NzSkeletonParagraph {
+    const basicProps: NzSkeletonParagraph = {};
+    // Width
+    if (hasAvatar && hasTitle) {
+      basicProps.width = '100%';
+    } else {
+      basicProps.width = '61%';
+    }
+    // Rows
+    if (!hasAvatar && hasTitle) {
+      basicProps.rows = 3;
+    } else {
+      basicProps.rows = 2;
+    }
+    return basicProps;
+  }
+
+  getProps<T>(prop: T | boolean | undefined): T | {}  {
+    if (prop && typeof prop === 'object') {
+      return prop;
+    }
+    return {};
+  }
+
+  toCSSUnit(value: number | string = ''): string {
+    if (typeof value === 'number') {
+      return `${value}px`;
+    } else if (typeof value === 'string') {
+      return value;
+    }
+  }
+
+  getWidthList(): Array<number | string> {
+    const { width, rows } = this.paragraph;
+    let widthList = [];
+    if (width && Array.isArray(width)) {
+      widthList = width;
+    } else if (width && !Array.isArray(width)) {
+      widthList = [];
+      widthList[rows - 1] = width;
+    }
+    return widthList;
+  }
+
+  updateClassMap(): void {
+    this.avatarClassMap = {
+      [ `ant-skeleton-avatar-lg` ]     : this.avatar.size === 'large',
+      [ `ant-skeleton-avatar-sm ` ]    : this.avatar.size === 'small',
+      [ `ant-skeleton-avatar-circle` ] : this.avatar.shape === 'circle',
+      [ `ant-skeleton-avatar-square ` ]: this.avatar.shape === 'square'
+    };
+  }
+
+  updateProps(): void {
+    this.title = {
```

Comment:
I know you're trying to cover default props with user's customized props. But these `...` operators seems verbose to me. Could you figure out a simpler and direct way to prepare these parameters?